Remove Taproot BIP 9 deployment

Sjors opened the pull request:
Drop `DEPLOYMENT_TAPROOT` from `consensus.vDeployments`.

Now that the commit (https://github.com/bitcoin/bitcoin/commit/7c08d81e119570792648fe95bbacddbb1d5f9ae2) which changes taproot to be enforced for all blocks is included in v24.0, it seems a good time to remove no longer needed non-consensus code.

Clarify what is considered a `BuriedDeployment`.

We drop taproot from `getdeploymentinfo` RPC, rather than mark it as `buried`, because this is not a buried deployment in the sense of [BIP 90](https://github.com/bitcoin/bips/blob/master/bip-0090.mediawiki). This is because the activation height has been completely removed from the code, unlike the hardcoded `DEPLOYMENT_SEGWIT` height which is still relied on.[^1]

See discussion in #24737 and #26162.

[^1]: `DEPLOYMENT_SEGWIT` is used in `IsBlockMutated` to determine if witness data is allowed, it's used for [BIP147](https://github.com/bitcoin/bips/blob/master/bip-0147.mediawiki), to trigger `NeedsRedownload()` for users who upgraded after a decade, and for a few other things.

luke-jr commented:
Concept ~0, but I object the claim in general that commits "sufficiently buried by other commits, and thus less likely to be reverted". Git isn't PoW, and "burying" commits does not make them less likely to be reverted.

Sjors commented:
@luke-jr changed the wording to "is (probably) included in v24.0".

ajtowns commented:
This should be updating `MinBIP9WarningHeight` above the taproot activation height (otherwise you should see `"warnings": "Unknown new rules activated (versionbit 2)"` due to miner signalling)

Sjors commented:
@ajtowns done

MarcoFalke commented on src/consensus/params.h:
```diff
@@ -94,7 +96,7 @@ struct Params {
     int SegwitHeight;
     /** Don't warn about unknown BIP 9 activations below this height.
-     * This prevents us from warning about the CSV and segwit activations. */
+     * This prevents us from warning about the CSV,segwit and taproot activations. */
```
```suggestion
     * This prevents us from warning about the CSV, segwit and taproot activations. */
```

nit

MarcoFalke commented on doc/release-notes-26201.md:
```diff
@@ -0,0 +1,5 @@
+RPC
+---
+
+- 'taproot' has been removed from 'getdeploymentinfo', rather than marked as buried,
```
```suggestion
- 'taproot' has been removed from 'getdeploymentinfo',
```

nit: I think release notes should describe changes, not provide extended discussion about alternatives

achow101 commented:
Are you still working on this?

Sjors commented:
I'll rebase soon(tm).

luke-jr commented on src/deploymentinfo.cpp:
```diff
@@ -11,10 +11,6 @@ const struct VBDeploymentInfo VersionBitsDeploymentInfo[Consensus::MAX_VERSION_B
         /*.name =*/ "testdummy",
         /*.gbt_force =*/ true,
     },
-    {
-        /*.name =*/ "taproot",
-        /*.gbt_force =*/ true,
-    },
```
This will need `aRules` in `rpc/mining.cpp` updated

Sjors commented on src/rpc/mining.cpp:
```diff
@@ -861,7 +861,10 @@ static RPCHelpMan getblocktemplate()
-    if (!fPreSegWit) aRules.push_back("!segwit");
+    if (!fPreSegWit) {
+        aRules.push_back("!segwit");
+        aRules.push_back("!taproot");
```
@luke-jr like this?

> This will need `aRules` in `rpc/mining.cpp` updated

Sjors replied:
Oh wait, without `!`

I'll push a fix and add a test...

Before and after this change it should be:

```
bitcoin-cli getblocktemplate '{"rules": ["segwit"]}' | jq '.rules'
[
  "csv",
  "!segwit",
  "taproot"
]
```

Sjors commented:
Rebased after kernel changes. Fixed a regression in `getblocktemplate`'s `rules` result, and added a test for tit.

Sjors commented on src/rpc/mining.cpp:
```diff
@@ -860,8 +860,14 @@ static RPCHelpMan getblocktemplate()
+    // Otherwise clients may assume the rule will not impact usage of the template as-is.
     aRules.push_back("csv");
-    if (!fPreSegWit) aRules.push_back("!segwit");
+    if (!fPreSegWit) {
```
In #27433 I propose getting rid of `fPreSegWit`.

Sjors commented:
Rebased after #28591.

Sjors commented on src/consensus/params.h:
```diff
@@ -18,20 +18,22 @@ namespace Consensus {
 
 enum DeploymentPos : uint16_t {
     DEPLOYMENT_TESTDUMMY,
-    DEPLOYMENT_TAPROOT, // Deployment of Schnorr/Taproot (BIPs 340-342)
```
Added. IIUC the WarningBitsConditionChecker uses ComputeBlockVersion (also used by the miner), which in turn uses VersionBitsConditionChecker, which uses vDeployments to determine if a softfork is STARTED or LOCKED_IN. If so it won't issue the warning. This logic is rather convoluted…

Sjors commented:
Rebased because #27433 needed it.

Sjors commented:
Rebased due to (trivial) merge conflict after #29547.

Sjors commented:
Rebased after #30560 and #29775.

Sjors commented:
Rebased after #30658.

Sjors commented:
#27433 needed the rebase here

Sjors commented:
Removed ~ from the description (wasn't relevant anymore anyway), since @jonatack pointed out in #28358 that these don't get rendered nicely in merges.

l0rinc reviewed (approved):
Concept ACK

Sjors commented:
Rebased after #31978.

jonatack reviewed (commented):
Light approach ACK 301e41985ef8cc8e3098d1d737864a5ae1f5108e

jonatack commented on src/consensus/params.h:
```diff
@@ -18,21 +18,24 @@ namespace Consensus {
 /**
  * A buried deployment is one where the height of the activation has been hardcoded into
  * the client implementation long after the consensus change has activated. See BIP 90.
+ * Consensus changes for which the new rules enforced from genesis are not listed here.
```
Should this be like https://github.com/bitcoin/bitcoin/pull/26201/files#diff-decae4be02fb8a47ab4557fe74a9cb853bdfa3ec0fa1b515c0a1e5de91f4ad0bR1418?

```suggestion
 * Consensus changes for which the new rules are enforced from genesis are not listed here.
```

Sjors replied:
Done

Sjors commented:
Rebased after #29039.

Sjors commented:
Trivial rebase after #32386.

Sjors commented:
Rebased after https://github.com/bitcoin/bitcoin/pull/31981 for minor conflict in `mining_basic.py`.

Sjors commented:
Rebased after splitting the `doc/bips.md` changes off into https://github.com/bitcoin/bitcoin/pull/32776.

Sjors commented:
Rebased after #33274.

darosior reviewed (commented):
Concept ACK.

I find it confusing that we don't simply make it a buried deployment like we did for the past 5 soft forks. But since we always enforce it from genesis, it would probably be even more confusing (for instance setting `-testactivationheight=taproot@42` wouldn't have the intended effect).

Could you update the PR title? There is no Taproot activation height. Maybe "Remove Taproot BIP 9 deployment"?

ajtowns commented on src/rpc/blockchain.cpp:
```diff
@@ -1437,15 +1437,15 @@ UniValue DeploymentInfo(const CBlockIndex* blockindex, const ChainstateManager&
     return RPCHelpMan{"getdeploymentinfo",
-        "Returns an object containing various state info regarding deployments of consensus changes.",
+        "Returns an object containing various state info regarding deployments of consensus changes.\n"
+        "Consensus changes for which the new rules are enforced from genesis are not listed here.",
```
Since #32998 we report `"script_flags": [ .., "TAPROOT", .. ]` via getdeploymentinfo for all blocks except the hardcoded exceptions. I think that's sufficient indication that the soft-fork is still enforced even if it's not listed in the "deployments" section anymore. Maybe this comment would be better phrased as 'are not listed ~~here~~ in "deployments".' ?

ajtowns reviewed (commented):
ACK ec17c82846686cb56da0535051bb701b3c84cfea

sedited reviewed (approved):
ACK ec17c82846686cb56da0535051bb701b3c84cfea

Sjors committed:
Remove Taproot activation height

Drop DEPLOYMENT_TAPROOT from consensus.vDeployments.

Bump MinBIP9WarningHeight.

Clarify what is considered a BuriedDeployment and
drop taproot from getdeploymentinfo RPC.

Add a test to getblocktemplate to ensure the taproot
rule is still set.

Co-Authored-By: MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>

sedited reviewed (approved):
Re-ACK 74f71c5054ff32a9ba922eed59fceeaf79ec1ed8

darosior reviewed (approved):
utACK 74f71c5054ff32a9ba922eed59fceeaf79ec1ed8

fanquake commented:
Note that this resulted in atleast some changes downstream: https://github.com/lightningnetwork/lnd/pull/10683

Sjors commented:
I opened #35183 to document this.
